Cassandra 15861 lock trunk

jasonstack opened the pull request.

jasonstack committed:
Test ONLY - DO NOT MERGE

jasonstack committed:
Guard entire-sstable-streaming ComponentManifest creation with read lock to avoid racing with STATS mutation and index summary redistribution.

* Lock is needed to prevent manifest linking partially written components.
* Hard links are used to avoid original component being deleted or replaced.
* Possible blocking: saving index summary or stats metadata may block manifest creation, but they are not very frequent.

maedhroz commented:
@jasonstack FWIW, both of the dtests that failed [here](https://app.circleci.com/pipelines/github/jasonstack/cassandra/244/workflows/9e2af3a1-7b63-423d-8cde-d2cd178c81d6/jobs/2329) passed locally for me.

maedhroz commented on test/unit/org/apache/cassandra/db/streaming/CassandraEntireSSTableStreamWriterTest.java:
```diff
@@ -126,43 +152,98 @@ public void testBlockWriterOverWire() throws IOException
+            TableMetadata origin = store.metadata();
+            MigrationManager.announceTableUpdate(origin.unbuild().minIndexInterval(1).maxIndexInterval(2).build(), true);
+
+            try (LifecycleTransaction txn = store.getTracker().tryModify(asList(sstable), OperationType.UNKNOWN))
```
Should `UNKNOWN` be `INDEX_SUMMARY`?

jasonstack replied:
+1

maedhroz commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -1199,7 +1219,15 @@ else if (samplingLevel < indexSummary.getSamplingLevel())
+            tidy.global.writeLock.lock();
+            try
+            {
+                saveSummary(newSummary);
```
@jasonstack It looks like this is the only summary saving path that could affect streaming, but wanted to make sure...

jasonstack replied:
yeah, this is the only one. I think the other index-summary modifications are used before being part of the node.

maedhroz commented on test/unit/org/apache/cassandra/db/streaming/CassandraEntireSSTableStreamWriterTest.java:
```diff
@@ -126,43 +152,98 @@ public void testBlockWriterOverWire() throws IOException
+                                                                                         nonRedistributingOffHeapSize,
+                                                                                         memoryPoolBytes));
+            }
+            descriptor.getMetadataSerializer().mutateRepairMetadata(descriptor, 0, UUID.randomUUID(), false);
```
@jasonstack Do we need to mutate the repair metadata if `redistributeSummaries()` saves the index summary, which is what we seem to be testing in this case?

jasonstack replied:
good catch.. will remove it

maedhroz commented on test/unit/org/apache/cassandra/db/streaming/CassandraEntireSSTableStreamWriterTest.java:
```diff
@@ -126,43 +152,98 @@ public void testBlockWriterOverWire() throws IOException
+            return mutation;
+        });
+
+        // wait for concurrent component mutation to finish
```
nit: For the mutation and the write to finish?

jasonstack replied:
`writer.write` is blocking, it's already finished.

maedhroz commented on test/unit/org/apache/cassandra/db/streaming/CassandraEntireSSTableStreamWriterTest.java:
```diff
@@ -126,43 +152,98 @@ public void testBlockWriterOverWire() throws IOException
+            {
+                mutation = ScheduledExecutors.optionalTasks.submit(componentMutation);
+                // wait 1s for concurrent task to start
+                FBUtilities.sleepQuietly(1000);
```
nit: Rather than a timed wait, what about just counting down on a latch when the provided `Callable` actually starts. (We'd probably have to make it a field in the test class, but it can be reset in a `@Before` block, etc.)

jasonstack replied:
+1

maedhroz commented on test/unit/org/apache/cassandra/db/streaming/CassandraEntireSSTableStreamWriterTest.java:
```diff
@@ -126,43 +152,98 @@ public void testBlockWriterOverWire() throws IOException
+        StreamMessageHeader messageHeader = new StreamMessageHeader(sstable.metadata().id, peer, session.planId(), false, 0, 0, 0, null);
+        CassandraEntireSSTableStreamReader reader = new CassandraEntireSSTableStreamReader(messageHeader, header, session);
 
         SSTableMultiWriter sstableWriter = reader.read(new DataInputBuffer(serializedFile.nioBuffer(), false));
```
nit: Should we instantiate `sstableWriter` at the start of a try-with-resources block?

jasonstack replied:
good idea

maedhroz commented on test/unit/org/apache/cassandra/db/streaming/CassandraEntireSSTableStreamWriterTest.java:
```diff
@@ -105,14 +122,23 @@ public static void defineSchemaAndPrepareSSTable()
+        CassandraEntireSSTableStreamWriter writer = new CassandraEntireSSTableStreamWriter(sstable, session, manifest);
 
         EmbeddedChannel channel = new EmbeddedChannel();
         AsyncStreamingOutputPlus out = new AsyncStreamingOutputPlus(channel);
```
Minor nit on the `Queue` below...could be `Queue<?>` to avoid a compiler warning.

jasonstack replied:
+1

maedhroz commented on test/unit/org/apache/cassandra/db/streaming/CassandraEntireSSTableStreamWriterTest.java:
```diff
@@ -126,43 +152,98 @@ public void testBlockWriterOverWire() throws IOException
+            }
+
+            EmbeddedChannel channel = createMockNettyChannel(serializedFile);
+            AsyncStreamingOutputPlus out = new AsyncStreamingOutputPlus(channel);
```
nit: Do we need to close the channel/output?

jasonstack replied:
added try-with-resources

maedhroz commented on src/java/org/apache/cassandra/db/compaction/CompactionStrategyManager.java:
```diff
@@ -1200,8 +1200,7 @@ public void mutateRepaired(Collection<SSTableReader> sstables, long repairedAt,
             {
-                sstable.descriptor.getMetadataSerializer().mutateRepairMetadata(sstable.descriptor, repairedAt, pendingRepair, isTransient);
-                sstable.reloadSSTableMetadata();
+                sstable.mutateAndReloadStats(stats -> stats.mutateRepairedMetadata(repairedAt, pendingRepair, isTransient));
```
@jasonstack Do we lose trace recording if we call `mutate()` directly rather than `IMetadataSerializer#mutateRepairMetadata()`?

jasonstack replied:
I can put the trace back.. or revert the change if stats component is preloaded.

maedhroz replied:
@jasonstack Ah, just a reminder to put this back if possible before we commit, given we went the hard-link route.

jasonstack replied:
added an `description` argument to pass trace info

maedhroz commented on src/java/org/apache/cassandra/db/compaction/LeveledManifest.java:
```diff
@@ -164,8 +164,7 @@ public synchronized void add(SSTableReader reader)
                 logger.debug("Could not add sstable {} in level {} - dropping to 0", reader, reader.getSSTableLevel());
-                reader.descriptor.getMetadataSerializer().mutateLevel(reader.descriptor, 0);
-                reader.reloadSSTableMetadata();
+                reader.mutateAndReloadStats(stats -> stats.mutateLevel(0));
```
Same as above. Does this mean we no longer trace like we would have in `IMetadataSerializer#mutateLevel()`?

jasonstack replied:
reverted to preload stats component

maedhroz commented on src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java:
```diff
@@ -127,6 +135,8 @@ public String getName()
     @Override
     public long getSize()
```
nit: Perhaps we should just rename it `getEstimatedSize()`?

jasonstack replied:
+1

maedhroz commented on src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java:
```diff
@@ -127,6 +135,8 @@ public String getName()
     {
+        // File size may not be accurate due to concurrent stats update or index summary redistribution. Since it's only
+        // used for metrics, an estimation should be fine.
         return header.size();
```
If this doesn't happen much, what about just generating the header on the fly? We'd at least avoid the slight confusion of having both an instance field and the `current` header for entire-SSTable streaming...

jasonstack replied:
I didn't compute size on the fly because the size computation is slightly slow for non-entire-streaming with compression..

maedhroz commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -1199,7 +1219,15 @@ else if (samplingLevel < indexSummary.getSamplingLevel())
+            finally
+            {
+                tidy.global.writeLock.unlock();
+            }
```
What do you think about having a `runWithWriteLock(Runnable)` or something similar on `GlobalTidy`? I'm guessing we're not concerned about the overhead of crating a lambda, etc.

jasonstack replied:
+1

maedhroz reviewed (commented):
@jasonstack Made a first pass and dropped a few (mostly minor) comments. I think the big thing to resolve is the discussion [here](https://issues.apache.org/jira/browse/CASSANDRA-15861?focusedCommentId=17162356&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17162356).

Thanks for the patch!

jasonstack commented on src/java/org/apache/cassandra/db/streaming/CassandraStreamHeader.java:
```diff
@@ -119,6 +119,11 @@ private long calculateSize()
                 transferSize += chunk.length + 4; // 4 bytes for CRC
         }
+        // Delay the creation of compressionInfo at sender to reduce GC pressure, see CASSANDRA-10680.
+        else if (compressionMetadata != null)
```
this was accidentally removed in storage engine refactor..

maedhroz commented on src/java/org/apache/cassandra/db/streaming/ComponentManifest.java:
```diff
@@ -30,16 +35,31 @@
+
+import static java.lang.Math.min;
 
 public final class ComponentManifest implements Iterable<Component>
```
We could almost call this thing `ComponentStreamer` now :)

(I mean even if we go the hard-link + atomic move direction, it could be nice to delegate that here, given we'll have a spec per component that includes either the hard link or the canonical file to stream.)

jasonstack replied:
I have moved the streaming code out of `ComponentManifest`..

maedhroz commented on src/java/org/apache/cassandra/streaming/OutgoingStream.java:
```diff
@@ -47,7 +47,12 @@
-    long getSize();
+
+    /**
+     * @return estimated file size to be streamed. This should only be used for metrics only, because concurrent
```
grammar pedant: Could get rid of the second/redundant "only"

jasonstack replied:
+1

maedhroz commented on src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java:
```diff
@@ -55,55 +56,79 @@
 
     @VisibleForTesting
-    public static ComponentManifest getComponentManifest(SSTableReader sstable)
+    public static ComponentManifest getComponentManifest(SSTableReader sstable, boolean createHardLink)
```
@jasonstack What if we just moved `getComponentManifest()` to `ComponentManifest` itself as a static factory, maybe `ComponentManifest.create()`, and centralized all manifest creation there? (Might have to move `STREAM_COMPONENTS` there as well?)

jasonstack replied:
+1

maedhroz commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -2067,6 +2078,25 @@ public int getSSTableLevel()
+        finally
+        {
+            tidy.global.writeLock.unlock();
+        }
```
@jasonstack Let's move this block to `runWithWriteLock()`. It might help future readers looking at usages of `runWithWriteLock()` trace back to the stats writing code itself, rather than having to search on usages of the write lock itself.

jasonstack replied:
+1, missed it when reverting code..

maedhroz commented on test/unit/org/apache/cassandra/streaming/EntireSSTableStreamingCorrectFilesCountTest.java:
```diff
@@ -127,7 +127,7 @@ public void test() throws Exception
         int totalNumberOfFiles = session.transfers.get(store.metadata.id).getTotalNumberOfFiles();
 
-        assertEquals(CassandraOutgoingFile.getComponentManifest(sstable).components().size(), totalNumberOfFiles);
+        assertEquals(CassandraOutgoingFile.getComponentManifest(sstable, false).components().size(), totalNumberOfFiles);
```
@jasonstack Would it be useful to have a simple test that ensures hard links get cleaned up when streaming completes? (perhaps an additional assertion in one of the other tests...)

jasonstack replied:
+1

maedhroz commented on src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java:
```diff
@@ -154,29 +179,43 @@ public UUID getPendingRepair()
+            // file sizes. (Note: Windows doesn't support atomic replace and index summary redistribution deletes
+            // existing file first)
+            // Recreate the latest manifest with hard links in case components are modified.
+            ComponentManifest manifest = sstable.runWithReadLock(ignored -> getComponentManifest(sstable, true));
```
@jasonstack If we only hold the read lock for manifest creation, and we _don't_ create a hard link for the index summary component, what happens if, after we create the manifest, we start streaming, and then an index summary redistribution finishes and tries to save? It seems like creating a hard link for the index summary would work here, even if we don't do an atomic replace, given manifest creation would have to block if a summary saving is in progress anyway.

jasonstack replied:
so the index summary hark link is created on either old summary or new summary, never on the partially written file because of the lock..

jasonstack replied:
ah.. forgot to hard link `index summary`..

maedhroz commented on test/unit/org/apache/cassandra/db/streaming/CassandraEntireSSTableStreamWriterTest.java:
```diff
@@ -105,69 +124,144 @@ public static void defineSchemaAndPrepareSSTable()
+            {
+                mutation = ScheduledExecutors.optionalTasks.submit(componentMutation);
+                // wait for concurrent task to start
+                waitForLatch(10, TimeUnit.SECONDS);
```
@jasonstack Out of curiosity, I disabled all the read/write lock acquisition in this patch, and the tests here all seem to pass pretty consistently still. I think what we might need to do is take the concept here, and most of the code, and test w/ two threads, one streaming at the `CassandraOutgoingFile#write()` level, where our read locking actually occurs, and the other guaranteed to make a mutation to the stats or index summary between initial manifest creation and streaming. (i.e. Something that would clearly fail before this patch.)

I think the other thing that made this more apparent to me after looking a second time was the fact that we actually call `runWithReadLock()` explicitly in `testBlockReadingAndWritingOverWire()`. We should be testing the locking mechanism in a more black box fashion, not replacing part of that mechanism in the test itself.

WDYT?

(I don't know if we should literally move the tests to `CassandraOutgoingFileTest`, but `CassandraEntireSSTableStreamWriter` is clearly not where we added the new logic.)

jasonstack replied:
agreed, we need another test to verify the lock. These tests in `CassandraEntireSSTableStreamWriter` only verify hard links are working as expected (`hard links` are not impacted by replace/deletion).

jasonstack replied:
agreed, we need some tests for locks. `CassandraEntireSSTableStreamWriter` only verifies hard links are working as expected, ie. `not affected by file deletion or replacement`

maedhroz commented on src/java/org/apache/cassandra/db/streaming/ComponentManifest.java:
```diff
@@ -18,28 +18,85 @@
+                                                                             Component.COMPRESSION_INFO, Component.FILTER, Component.SUMMARY,
+                                                                             Component.DIGEST, Component.CRC);
+
+    private static final Set<Component> MUTABLED_COMPONENTS = ImmutableSet.of(Component.STATS, Component.SUMMARY);
```
nit: `MUTABLED_COMPONENTS` -> `MUTABLE_COMPONENTS`

maedhroz commented on test/unit/org/apache/cassandra/db/streaming/CassandraEntireSSTableStreamWriterTest.java:
```diff
@@ -112,7 +111,7 @@ public void testBlockWriterOverWire() throws IOException
-        CassandraEntireSSTableStreamWriter writer = new CassandraEntireSSTableStreamWriter(sstable, session, CassandraOutgoingFile.getComponentManifest(sstable));
+        CassandraEntireSSTableStreamWriter writer = new CassandraEntireSSTableStreamWriter(sstable, session, ComponentManifest.create(sstable, false));
 
         EmbeddedChannel channel = new EmbeddedChannel();
```
@jasonstack Should we still close this channel? (It looks like we more or less reverted this back to what's in trunk, which is fine, but wanted to make sure we didn't throw the baby out with the bath water...)

jasonstack replied:
+1

maedhroz commented on test/unit/org/apache/cassandra/db/streaming/EntireSSTableStreamConcurrentComponentMutationTest.java:
```diff
@@ -0,0 +1,338 @@
+    private static void countDown()
+    {
+        latch.countDown();
+    }
```
formatting: missing newline?

jasonstack replied:
+1

maedhroz commented on test/unit/org/apache/cassandra/db/streaming/EntireSSTableStreamConcurrentComponentMutationTest.java:
```diff
@@ -0,0 +1,338 @@
+    @BMRule(name = "Let component mutation happen immediately after ComponentManifest completion",
+    targetClass = "ComponentManifest",
+    targetMethod = "<init>",
+    action = "Thread.sleep(3000)")
```
formatting: horizontal alignment w/ `name`?

jasonstack replied:
+1

maedhroz commented:
@jasonstack btw, there's a random unnecessary semicolon in `SSTableReader#saveSummary()`

maedhroz commented on test/unit/org/apache/cassandra/db/streaming/EntireSSTableStreamConcurrentComponentMutationTest.java:
```diff
@@ -0,0 +1,338 @@
+        SchemaLoader.createKeyspace(KEYSPACE,
+                                    KeyspaceParams.simple(1),
+                                    SchemaLoader.standardCFMD(KEYSPACE, CF_STANDARD),
+                                    SchemaLoader.compositeIndexCFMD(KEYSPACE, CF_INDEXED, true),
```
We don't actually need to load/create these other two column families for this test, right?

jasonstack replied:
+1

maedhroz commented on test/unit/org/apache/cassandra/db/streaming/EntireSSTableStreamConcurrentComponentMutationTest.java:
```diff
@@ -0,0 +1,338 @@
+        return true;
+    }
+
+    private Future<?> executeAsync(Callable<?> mutation)
```
nit: This isn't always a mutation.

jasonstack replied:
+1

maedhroz commented on test/unit/org/apache/cassandra/db/streaming/EntireSSTableStreamConcurrentComponentMutationTest.java:
```diff
@@ -0,0 +1,338 @@
+            targetMethod = "saveSummary(Descriptor, DecoratedKey, DecoratedKey, IndexSummary)",
+            targetLocation = "AFTER INVOKE serialize",
+            condition = "$descriptor.cfname.contains(\"Standard1\")",
+            action = "org.apache.cassandra.db.streaming.EntireSSTableStreamConcurrentComponentMutationTest.countDown();Thread.sleep(5000);")
```
@jasonstack Awesome. I think this is going to work in most environments. (It certainly works locally for me.) What if, instead of relying on a 5 second wait, we just wait on a separate latch, sort of the opposite of what we do with `countDown()` immediately before? Then we'll suspend the summary saving from finishing while the write continues. When the write is done, we can have it count down on that second latch and free the other thread to finish the save. WDYT?

(I know we've both been bitten by slow testing environments messing with our timing assumptions in the past.)

jasonstack replied:
>  Then we'll suspend the summary saving from finishing while the write continues. 

It does reproduce the issue more reliably when there is no lock.. With lock, the test will run into deadlock because index summary didn't release lock, manifest cannot be created..

maedhroz commented on test/unit/org/apache/cassandra/db/streaming/EntireSSTableStreamConcurrentComponentMutationTest.java:
```diff
@@ -0,0 +1,338 @@
+    action = "Thread.sleep(3000)")
+    public void testStreamWithStatsMutation() throws Exception
+    {
+        testStreamWithConcurrentComponentMutation(NO_OP, () -> {
```
@jasonstack Crazy idea: What if we just did the mutation as the `runBeforeStreaming` callable here? We could at least avoid the hard-coded sleep (although the `mutation` callable might need a more generalized name, like `runConcurrentWithStreaming` or something).

jasonstack replied:
+1, this will prove the need of on-demand manifest creation, rather than relying on the manifest created initially.

maedhroz commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -2067,6 +2081,21 @@ public int getSSTableLevel()
+     * @param description on changed attributes
+     * @param transform function to mutate stats metadata
+     */
+    public void mutateAndReloadStats(String description, UnaryOperator<StatsMetadata> transform) throws IOException
```
So I guess there's one other thing you could do that would again simplify the tracing stuff we added back, and that's breaking `mutateAndReloadStats()` back into a few different methods. (i.e. One for changing repairedAt time, one for level changes, etc.) It might at least make piecing together the actual trace message easier. The `runWithWriteLock()` wrapping isn't really that much boilerplate to duplicate among them for the sake of clarity.

jasonstack replied:
good idea.

maedhroz commented on src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java:
```diff
@@ -154,29 +143,43 @@ public UUID getPendingRepair()
+            // file sizes. (Note: Windows doesn't support atomic replace and index summary redistribution deletes
+            // existing file first)
+            // Recreate the latest manifest with hard links in case components are modified.
+            ComponentManifest manifest = sstable.runWithReadLock(ignored -> ComponentManifest.create(sstable, true));
```
I suppose we could make `ComponentManifest` `AutoCloseable`. (I guess it wouldn't be necessary to close in all cases though...)

jasonstack replied:
+1 Added `ComponentContext`

maedhroz commented on src/java/org/apache/cassandra/db/streaming/ComponentManifest.java:
```diff
@@ -63,6 +120,12 @@ public long totalSize()
     }
 
+    @VisibleForTesting
+    public Collection<File> hardlinks()
```
@jasonstack Is this unused now?

jasonstack replied:
removed

jasonstack committed:
Use CompontContext to create hard links and channels on write side

jasonstack committed:
Used synchronized instead of reentrant lock

maedhroz commented on src/java/org/apache/cassandra/db/streaming/ComponentContext.java:
```diff
@@ -0,0 +1,94 @@
+    /**
+     * @return file channel to be streamed, either original component or hardlinked component.
+     */
+    @SuppressWarnings("resource") // file channel will be closed by Caller
```
@jasonstack I think what we actually want is a `@SuppressWarnings("IOResourceOpenedButNotSafelyClosed")` on the line where we create the `RandomAccessFile`. (Then we can remove this one.)

jasonstack replied:
moved `@SuppressWarnings ` to `RandomAccessFile `, but using `resource` seems cleaner than `IOResourceOpenedButNotSafelyClosed`..

maedhroz replied:
Does "resource" actually avoid the warning?

jasonstack replied:
`ant eclipse-warnings` didn't complain..

maedhroz replied:
Odd. Perhaps it was just an IDEA issue.

maedhroz commented on src/java/org/apache/cassandra/db/streaming/ComponentContext.java:
```diff
@@ -0,0 +1,94 @@
+    @Override
+    public void close()
+    {
+        hardLinks.values().forEach(File::delete);
```
@jasonstack Should we catch and do something with exceptions that might be thrown from `delete()`? I suppose we could use `FileUtils.deleteWithConfirm()`, then `clear()`, then log any exceptions at warn?

jasonstack replied:
+1

maedhroz commented on test/microbench/org/apache/cassandra/test/microbench/ZeroCopyStreamingBenchmark.java:
```diff
@@ -40,12 +40,7 @@
-import org.apache.cassandra.db.streaming.CassandraStreamHeader;
-import org.apache.cassandra.db.streaming.CassandraStreamReader;
-import org.apache.cassandra.db.streaming.CassandraStreamWriter;
+import org.apache.cassandra.db.streaming.*;
```
nit: I think we generally try to avoid wildcards.

jasonstack replied:
+1

maedhroz commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -2067,6 +2079,30 @@ public int getSSTableLevel()
+        {
+            descriptor.getMetadataSerializer().mutateLevel(descriptor, newLevel);
+            reloadSSTableMetadata();
+        };
```
nit: remove `;`

jasonstack replied:
+1

maedhroz commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -2067,6 +2079,30 @@ public int getSSTableLevel()
+        {
+            descriptor.getMetadataSerializer().mutateRepairMetadata(descriptor, newRepairedAt, newPendingRepair, isTransient);
+            reloadSSTableMetadata();
+        };
```
nit: remove `;`

jasonstack replied:
+1

maedhroz commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -1202,39 +1215,38 @@ public void run()
-            double effectiveInterval = indexSummary.getEffectiveIndexInterval();
+        int minIndexInterval = metadata().params.minIndexInterval;
+        int maxIndexInterval = metadata().params.maxIndexInterval;
+        double effectiveInterval = indexSummary.getEffectiveIndexInterval();
```
@bdeggleston @jasonstack I [mentioned](https://issues.apache.org/jira/browse/CASSANDRA-15861?focusedCommentId=17170484&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17170484) this briefly in the Jira, but with `indexSummary` not `final` or `volatile` in `SSTableReader`, do we have a visibility issue? The thread that loads/builds the summary needs to safely publish its changes for the thread that builds new summaries here.

Simply making `indexSummary` `volatile` might work. Most of its accesses are themselves part of the summary redistribution task itself or during compaction, so we probably don't care about that overhead. `SSTableReader#getIndexScanPosition()`, on the other hand, does look like it's used at least during some range reads.

We could go back to the old scope for the `synchronized` block, leaving `indexSummary` non-volatile. That would at least ensure visibility between the `cloneX()` methods. Even then, there are other things that call `SSTableReader#internalOpen()`, like `BigTableWriter.openEarly()` and `openFinal()`, that don't initialize `indexSummary` while holding the global lock. (Perhaps we could just synchronize to get the `IndexSummary` reference, rather than having to wait for `buildSummaryAtLevel()`, though.)

Is there a bug here independent of this patch? i.e. Do we actually need to make sure the block that includes setting `indexSummary` in `SSTableReader#internalOpen()` is performed in a `synchronized (tidy.global)` block? We can't synchronize on `tidy`, given we create a new one for a new reader, but we should be able to sync on the right `GlobalTidy` even before set call `setup()`...or we could make `global` final and initialize it at `InstanceTidier` construction.

I hope I'm just missing something...

CC @belliottsmith

maedhroz replied:
There are other places that write `indexSummary` without synchronization as well, namely `SSTableReader#load()` and `openForBatch()`.

belliottsmith replied:
I would have to spend a while familiarising myself with this patch (and all present usages) to be sure, and all I've done to answer is skim the existing 3.0 SSTableReader, but this is probably fine. Any plain write that occurs before the SSTable is published will piggy-back on the `volatile` write used to publish the SSTable to other threads. I think in any cases we care about it _should_ piggy-back on `Tidy.setup` before this.  

The only likely issue, then, is if we were to modify the summary while the sstable is already published.  Even in this case we're fine because `IndexSummary` is an immutable class (by which I mean its fields are all final), so the worst that can happen is that it is not immediately visible (technically the JMM doesn't guarantee it is ever visible to other threads, but in practice it cannot fail to eventually make it visible)

maedhroz replied:
There _is_ a volatile write to `setup` in `InstanceTidier#setup()` after `ifile`, `indexSummary`, et al. are set on the `SSTableReader`, but I'm not sure that helps us reason about the visibility of `indexSummary` for the thread that enters `cloneWithNewSummarySamplingLevel()`?

Unless there's some consequence I'm missing around making `global` final and assigning it at `InstanceTidier` construction (which would at least get the static analysis tools off our backs), I'd really like to make our synchronization policy around the non-final fields of the reader explicit (i.e. around `tidy.global`) rather than relying on more subtle reasoning around a volatile.

In terms of how that all affects this patch, the fact that `IndexSummary` is immutable should make it possible for us to synchronize only to get the `IndexSummary` reference, rather than having to wait for `buildSummaryAtLevel()`.

belliottsmith replied:
So, for the result of `cloneWithNewSummaryAtLevel` to ever be visible to anybody, it must be published safely - in this case ordinarily through the DataTracker.View. So anything we do to provide visibility earlier than this is just providing is some nicer feels, which is fine, but as it happens we have several such volatile writes: the synchronised block when saving the new summary provides visibility, as does the write to `tidy.setup` (which is literally its purpose, iirc, though for the scope of `tidy`).

I am generally in favour of making more explicit links between a visibility requirement and where it is enforced, but there's definitely no need to introduce extra synchronised blocks - a simple write to a volatile property is sufficient, and in this case a comment indicating we're piggy-backing on any of these mentioned volatile writes might be sufficient - as would none of them, since we must publish the sstable safely anyway.  Stylistically, it's all up to you guys of course.

belliottsmith replied:
(This class is definitely a mess)

maedhroz replied:
I've had a chance to talk it over w/ @belliottsmith and think about it, and tl;dr the changes here should be safe. The big thing I was missing is that the `SSTableReader` itself should already be published safely between threads, and if it is, the `indexSummary` reference should be fine (i.e. not null, which is the only danger, because it doesn't actually change after creation given we clone).

However, it might be helpful to briefly document this in `cloneWithNewSummarySamplingLevel()` so future readers might avoid some confusion. Better yet, we could finally just make `indexSummary` final in `SSTableReader`, avoiding all this confusion in the first place. I'm guessing the reason they aren't is that we added some layers of initialization over time, but deferring reader creation until we're ready seems like it could be achieved with a builder (or even without one).

CC @jasonstack

belliottsmith replied:
+1, that would be a big improvement

jasonstack replied:
I made some refactor based on @maedhroz 's builder prototype: https://github.com/jasonstack/cassandra/pull/16. Now we have `BuilderForWriter`/`BuilderForRead`/`BuilderForBatch` and `IndexSummary` is `final`. We can make `dfile/ifile/bf` final as well, but didn't do it in this commit..

what do you guys think?

maedhroz replied:
@jasonstack I took a look and dropped a couple nits. Overall, it looks pretty straightforward, a largely organizational change that isolates building logic from the reader itself. I _do_ think that we should make `dfile` / `ifile` / `bf` final while we're here and have the opportunity. `first` and `last` seem a little more complicated, given the way the writer uses them during append, so no big deal if those stay mutable.

jasonstack replied:
updated the PR to make `dfile/ifile/bf` final.. `first/last` are used in `SSTable` and modified by writer.. I will leave them as they are..

maedhroz committed:
Use builder in SSTableReader to make indexSummary, dfile, ifile, bloomFilter final.

jasonstack committed:
apply caleb's feedback

jasonstack committed:
fix test

maedhroz commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -626,42 +575,25 @@ public void run()
+        builder.bf = bf;
+        builder.ifile = ifile;
+        builder.dfile = dfile;
+        builder.summary = summary;
```
nit: We could make a minor clean up here and have "chainable" methods for setting the bloom filter if you think that reads better.

ex.
```
BuilderForWriter builder =
    new BuilderForWriter(desc, metadata, maxDataAge).bloomFilter(bf)
                                                    .indexFile(ifile)
                                                    .dataFile(dfile)
                                                    .indexSummary(summary)
                                                    .build(components, sstableMetadata, openReason, header);
```

jasonstack replied:
good idea.

maedhroz commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -1125,18 +839,47 @@ private SSTableReader cloneAndReplace(DecoratedKey newFirst, OpenReason reason,
+        SSTableReader replacement = internalOpen(descriptor,
+                                                 components,
+                                                 metadata,
+                                                 ifile != null ? ifile.sharedCopy() : null,
```
@jasonstack Should `ifile` ever be `null` here? At least for this version of `cloneAndReplace()`, it looks like the only two things that use it (tests) pass without the null check...

jasonstack replied:
+1

maedhroz commented on src/java/org/apache/cassandra/io/sstable/format/big/BigFormat.java:
```diff
@@ -103,9 +104,11 @@ public SSTableWriter open(Descriptor descriptor,
-        public SSTableReader open(Descriptor descriptor, Set<Component> components, TableMetadataRef metadata, Long maxDataAge, StatsMetadata sstableMetadata, SSTableReader.OpenReason openReason, SerializationHeader header)
+        public SSTableReader open(Descriptor descriptor, Set<Component> components, TableMetadataRef metadata,
+                                  Long maxDataAge, StatsMetadata sstableMetadata, SSTableReader.OpenReason openReason,
+                                  SerializationHeader header, IndexSummary summary, FileHandle dfile, FileHandle ifile, IFilter bf)
```
nit: For now I guess we should at least put the arguments on separate lines.

jasonstack replied:
I extract `Builder` into a separate class `SSTableReaderBuilder` and pass it as the only argument to `open`.. WDYT?

maedhroz replied:
I had actually almost suggested that earlier. WFM

maedhroz commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableReader.java:
```diff
@@ -53,9 +55,11 @@
-    BigTableReader(Descriptor desc, Set<Component> components, TableMetadataRef metadata, Long maxDataAge, StatsMetadata sstableMetadata, OpenReason openReason, SerializationHeader header)
+    BigTableReader(Descriptor desc, Set<Component> components, TableMetadataRef metadata, Long maxDataAge,
+                   StatsMetadata sstableMetadata, OpenReason openReason, SerializationHeader header, IndexSummary summary,
+                   FileHandle dfile, FileHandle ifile, IFilter bf)
```
nit: newlines for arguments

jasonstack replied:
+1

maedhroz commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -430,52 +430,15 @@ public static SSTableReader openForBatch(Descriptor descriptor, Set<Component> c
-                                   OpenReason.NORMAL,
-                                   header.toHeader(metadata.get()));
+            BuilderForBatch builder = new BuilderForBatch(descriptor, metadata);
+            return builder.build(components, statsMetadata, OpenReason.NORMAL, header.toHeader(builder.metadata));
```
aside: It's kind of nice now that we don't have to even worry about creating a reader instance until the things that it depends on are actually ready/available.

maedhroz commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -701,16 +633,24 @@ protected SSTableReader(final Descriptor desc,
         selfRef = new Ref<>(this, tidy);
+        this.dfile = dfile;
+        this.ifile = ifile;
+        this.bf = bf;
```
OCD: Perhaps assign these together w/ the summary?

jasonstack replied:
+1

maedhroz commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -1125,18 +839,47 @@ private SSTableReader cloneAndReplace(DecoratedKey newFirst, OpenReason reason,
                                                  reason,
-                                                 header);
+                                                 header,
+                                                 newSummary);
```
nit: I suppose we could minimize the diff here by leaving the argument order (for `newSummary`) the same.

jasonstack replied:
+1

maedhroz commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -2430,7 +2187,10 @@ public abstract SSTableReader open(final Descriptor descriptor,
+                                           SerializationHeader header,
+                                           IndexSummary summary,
+                                           FileHandle dfile,
+                                           FileHandle ifile, IFilter bf);
```
nit: newline for `bf`

jasonstack replied:
+1

maedhroz commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -2538,4 +2298,369 @@ public static void shutdownBlocking(long timeout, TimeUnit unit) throws Interrup
+            }
+        }
+
+        public abstract SSTableReader build(Set<Component> components, StatsMetadata statsMetadata, OpenReason reason, SerializationHeader header);
```
Fun Fact: Everything would compile if we removed this abstract method ;)

jasonstack replied:
because it uses the concrete type..

maedhroz commented:
unrelated to patch nit: `GlobalTidy.NULL` should be `final`

maedhroz reviewed (approved):
@jasonstack I left one more round of nits, but otherwise, this patch LGTM

jasonstack committed:
Extract SSTableReaderBuilder into standalone class and pass builder as argument to Factory#open

jasonstack committed:
clean up

dcapwell commented on src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java:
```diff
@@ -154,29 +143,38 @@ public UUID getPendingRepair()
         {
-            CassandraEntireSSTableStreamWriter writer = new CassandraEntireSSTableStreamWriter(sstable, session, manifest);
-            writer.write((AsyncStreamingOutputPlus) out);
+            // Acquire lock to avoid concurrent sstable component mutation becaue of stats update or index summary
```
`s/becaue/because/`

jasonstack replied:
+1

dcapwell commented on src/java/org/apache/cassandra/db/streaming/ComponentContext.java:
```diff
@@ -0,0 +1,105 @@
+        hardLinks.clear();
+
+        if (accumulate != null)
+            logger.warn("Failed to remove hard link files: {}", accumulate.getMessage());
```
should be `logger.warn("Failed to remove hard link files", accumulate);`

jasonstack replied:
+1

dcapwell commented on src/java/org/apache/cassandra/db/streaming/ComponentContext.java:
```diff
@@ -0,0 +1,105 @@
+
+    public static ComponentContext create(Descriptor descriptor)
+    {
+        LinkedHashMap<Component, File> hardLinks = new LinkedHashMap<>(1);
```
any reason to use `LinkedHashMap` rather than `HashMap`?  this allocates a bit more memory and don't see a need for order (we delete files in the same order we created them, but don't think we require that property)

jasonstack replied:
reverted to hashmap

dcapwell reviewed (commented):
almost done with first pass, just need to review the builder changes as they moved a lot around.

So far it looks like we snapshot the two mutable components and use the global sstable lock to update the files and reload their contents.  Since the lock is shared, a slow write to SUMMARY (aka IndexSummaryRedistribution) will block updates to STATS (aka compaction) even though they do not logically conflict; streaming is only blocked while snapshotting.

jasonstack commented:
> a slow write to SUMMARY (aka IndexSummaryRedistribution) will block updates to STATS (aka compaction) even though they do not logically conflict; 

index summary redistribution and stats-mutation shouldn't happen together because both are running as part of compactions.

> streaming is only blocked while snapshotting.

streaming is also blocked by "write to summary file", as streaming should not snapshot a partially written file.

smiklosovic commented on src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java:
```diff
@@ -154,29 +143,38 @@ public UUID getPendingRepair()
     @Override
     public void write(StreamSession session, DataOutputStreamPlus out, int version) throws IOException
     {
+        boolean streamEntirely = shouldStreamEntireSSTable && out instanceof AsyncStreamingOutputPlus;
```
I believe this `out instaceof` is useless, if you check how that `out` is instantiated, it can never be something else.

dcapwell replied:
the API allows it to be other things, so would still require this check to avoid future changes causing class cast issues.

smiklosovic replied:
@maedhroz @dcapwell but then, suddenly, we are perfectly happy to cast that into `AsyncStreamingOutputPlus` without any checks and that write method also takes DataOutputStreamPlus ... Same for uncompressed / normal writer.
 https://github.com/apache/cassandra/blob/7542fd81abb445e10fbcc67929d742a9849b7129/src/java/org/apache/cassandra/db/streaming/CassandraCompressedStreamWriter.java#L61

dcapwell replied:
that seems problematic as `org.apache.cassandra.streaming.async.NettyStreamingMessageSender#sendControlMessage` calls this method (transitively) and passes in a `DataOutputBufferFixed`, which is not a `AsyncStreamingOutputPlus`.

I would need to look closer to see how they relate, but it does seem like this logic isn't safe.

Now, given the fact that this patch doesn't change this logic and the logic is already there; I am fine ignoring this fact in this JIRA, though that should be looked at closer as this seems like an edge case where `header.compressionInfo != null` might fail

blerer replied:
All the `CassandraStreamWriter` implementation will cast `DataOutputStreamPlus` into an `AsyncStreamingOutputPlus`.
In practice all `OutgoingStreamMessage` should be send through a `FileTask` which will alwas use an `AsyncStreamingOutputPlus`.
What we should have is an assert that the `DataOutputStreamPlus` is an `AsyncStreamingOutputPlus` as if it is not the case it is a programming error.

jasonstack replied:
> org.apache.cassandra.streaming.async.NettyStreamingMessageSender#sendControlMessage

This is for `KeepAliveMessage`..

For `OutgoingStreamMessage`, it must be `AsyncStreamingOutputPlus `

dcapwell commented:
> index summary redistribution and stats-mutation shouldn't happen together because both are running as part of compactions.

You are right, I keep forgetting `org.apache.cassandra.db.lifecycle.Tracker#tryModify(java.lang.Iterable<org.apache.cassandra.io.sstable.format.SSTableReader>, org.apache.cassandra.db.compaction.OperationType)` marks sstables as compacting, so compactions will avoid compacting those files

```
if (null == apply(permitCompacting(sstables), updateCompacting(emptySet(), sstables)))
            return null;
        return new LifecycleTransaction(this, operationType, sstables);
```

dcapwell commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReaderBuilder.java:
```diff
@@ -0,0 +1,467 @@
+
+                SSTableReader sstable = readerFactory.open(this);
+
+                sstable.first = first;
```
nit: would have loved if the builders made first/last immutable.  At the moment the read logic is now controlled via builder, the only complication is `org.apache.cassandra.io.sstable.format.big.BigTableWriter#afterAppend`.

this is unrelated to the zero copy work and locking, so fine to leave as is

maedhroz replied:
@dcapwell Yeah, we had some back and forth on this in one of the resolved comments above and ended in pretty much the same place.

jasonstack committed:
Add test on CassandraStreamHeader calculatedSize

jasonstack committed:
Apply blake's feedback

maedhroz commented on test/unit/org/apache/cassandra/db/streaming/CassandraStreamHeaderTest.java:
```diff
@@ -20,23 +20,148 @@
+    }
+
+    @Test
+    public void transferedSizeWithCompressionTest()
```
Verified that this fails without the fix in `calculateSize()`.

maedhroz replied:
CC @blerer @smiklosovic

jasonstack committed:
Apply David's feedback

blerer commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReaderBuilder.java:
```diff
@@ -0,0 +1,467 @@
+     * Load index summary from Summary.db file if it exists.
+     *
+     * if loaded index summary has different index interval from current value stored in schema,
+     * then Summary.db file will be deleted and this returns false to rebuild summary.
```
Nit: The original implementation was returning a boolean. The new one does not so the javadoc need to be fixed.
The code is also loading the `first` and `last` keys with the `indexSummary`. It might makes sense to change the method name and the documentation to make that more obvious.

jasonstack replied:
+1

blerer commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReaderBuilder.java:
```diff
@@ -0,0 +1,467 @@
+                    ByteBuffer key = ByteBufferUtil.readWithShortLength(primaryIndex);
+                    RowIndexEntry.Serializer.skip(primaryIndex, descriptor.version);
+                    DecoratedKey decoratedKey = metadata.partitioner.decorateKey(key);
+                    if (first == null)
+                        first = decoratedKey;
+                    last = decoratedKey;
```
I know that this code was the same before but does it make sense to recompute the `first` and `last` key if `summaryLoaded` is `true` as it means that they have been read from the `summary` file just before.

jasonstack replied:
updated, now guarded with `if (!summaryLoaded)`

blerer commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReaderBuilder.java:
```diff
@@ -0,0 +1,467 @@
+     * @param recreateBloomFilter true if recreate bloom filter
+     * @param summaryLoaded true if index summary is already loaded and not need to build again
+     */
+    void buildSummary(boolean recreateBloomFilter,
```
The javadoc and the method name are both misleading, as we are refactoring things it might be good to address that.

jasonstack replied:
+1

blerer commented on src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java:
```diff
@@ -154,29 +143,38 @@ public UUID getPendingRepair()
+            CassandraStreamHeader.serializer.serialize(header, out, version);
+            out.flush();
+
             CassandraStreamWriter writer = (header.compressionInfo == null) ?
```
The code around `CassandraStreamHeader.compressionInfo` is really misleading. On the sender side `compressinInfo` will be `null` until the header is serialized (the serialization calling `CassandraStreamHeader.calculateCompressionInfo`). Until that point the `CassandraStreamHeader.isCompressed` will return `false` even if the stream should be compressed. `equals` and `hashcode` are also affected by this effect.
For `CassandraStreamHeader.isCompressed` should check if the `compressionMetadata` is different from `null` and `compressionInfo`can be removed from `equals` and `hashcode` as they already use `compressionMetadata` and `sections` in their computation.
Regarding the `compressionInfo` field I wonder if it will not be better to simply make it as private and provide an accessor method that instantiate it lazily. It will be slightly less efficient but probably much safer. What do you think?

jasonstack replied:
> CassandraStreamHeader.isCompressed will return false even if the stream should be compressed.

`isCompressed()` is used on received side where `compressionMetadata` is always `null` regardless compressed. But good to make it consistent.

> Regarding the compressionInfo field I wonder if it will not be better to simply make it as private and provide an accessor method that instantiate it lazily.

Good idea, added accessor `getOrInitCompressionInfo`

> compressionInfo can be removed from equals and hashcode as they already use compressionMetadata and sections in their computation.

`equals`/`hashCode` didn't use `compressionMetadata`. Since `equals/hashCode` is not used in production code, I have changed them to use `getOrInitCompressionInfo()`, wdyt?

blerer replied:
I missed the receiving side logic with `compressionMetadata` vs `compressionInfo`. The changes you did will work, I just wonder if we cannot make the logic even less confusing. I tried to refactor things a bit and came up with:
https://github.com/apache/cassandra/commit/5c01763f7f8ae6b58a2d795101465d8195bf56d4
Tell me if it makes sense to you.

jasonstack replied:
Thanks for the refactoring, it looks great. I made a minor fix to use `Arrays.equals` for `chunks()` in `CompressionInfo#equals` https://github.com/apache/cassandra/pull/642/commits/7522f57e9397d94820a1020c1bba808b8a262e1a

blerer replied:
Good catch :-)

blerer commented on src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java:
```diff
@@ -154,29 +143,38 @@ public UUID getPendingRepair()
-                     new CassandraStreamWriter(sstable, header.sections, session) :
-                     new CassandraCompressedStreamWriter(sstable, header.sections,
-                                                         header.compressionInfo, session);
+                                           new CassandraStreamWriter(sstable, header.sections, session) :
+                                           new CassandraCompressedStreamWriter(sstable, header.sections, header.compressionInfo, session);
```
`CassandraStreamWriter` and `CassandraCompressedStreamWriter` reompute the size that was computed in `CassandraStreamHeader`. Rather than recomputing the size we should pass it from the header.

jasonstack replied:
good idea, updated to pass `CassandraStreamHeader` into both writers

jasonstack committed:
Apply feedback on SSTableReaderBuilder

jasonstack committed:
Apply feedback to CassandraOutgoingFile

jasonstack committed:
Apply feedback on CassandraStreamHefer.compressionInfo

blerer committed:
Refactor CompressionInfo and CassandraStreamHeader

The refactoring makes CompressionInfo abstract and provide 2 sub-classes.
One for the receiving side fully initialized and one for the sender side
that only computes the files chunks the first time that they are
requested.

The goal of the refactoring was to avoid the confusing
compressionMetadata/compressionInfo in CassandraStreamHeader.

jasonstack committed:
Use Arrays.equals for chunks()

maedhroz commented on test/microbench/org/apache/cassandra/test/microbench/ZeroCopyStreamingBenchmark.java:
```diff
@@ -149,23 +151,23 @@ public void setupBenchmark() throws IOException
             List<Range<Token>> requestedRanges = Arrays.asList(new Range<>(sstable.first.minValue().getToken(), sstable.last.getToken()));
-            partialStreamWriter = new CassandraStreamWriter(sstable, sstable.getPositionsForRanges(requestedRanges), session);
+            CassandraStreamHeader partialSSTableStreamHeader =
+            CassandraStreamHeader.builder()
```
nit: horizontal tab

jasonstack replied:
+1 updated format

jasonstack committed:
Apply format to ZeroCopyStreamingBenchmark
